zerver/lib/rate_limiter.py

```diff
@@ -160,6 +164,106 @@ def rate_limit_entity(cls, entity_key: str, rules: List[Tuple[int, int]],
         # Returns (ratelimited, secs_to_freedom)
         pass
 
+class TornadoInMemoryRateLimiterBackend(RateLimiterBackend):
+    # reset_times[key][rule] is the time at which the event
+    # request from the rate-limited key will be accepted.
+    reset_times = {}  # type: Dict[str, Dict[Tuple[int, int], float]]
+
+    # last_gc_time is the last time when the garbage was
+    # collected from reset_times.
+    last_gc_time = time.time()
+
+    # timestamps_blocked_until[key] contains the timestamp
+    # up to which the key has been blocked manually.
+    timestamps_blocked_until = {}  # type: Dict[str, float]
+
+    @classmethod
+    def need_to_limit(cls, entity_key: str, time_window: int,
+                      max_count: int) -> Tuple[bool, float]:
+        '''
+        Returns a tuple of `(rate_limited, time_till_free)`.
+        For simplicity, we have loosened the semantics here from
+        - each key may make atmost `count * (t / window)` request within any t
+          time interval.
+        to
+        - each key may make atmost `count * [(t / window) + 1]` request within
+          any t time interval.
+        Thus, we only need to store reset_times for each key which will be less
+        memory-intensive. This also has the advantage that you can only ever
+        lock yourself out completely for `window / count` seconds instead of
+        `window` seconds.
+        '''
+        now = time.time()
+
+        # Remove all timestamps from `reset_times` that are too old.
+        if cls.last_gc_time <= now - time_window / max_count:
```

Comment:
I'm a bit lost in this garbage cleaning block, can you explaing why this is how it should be?

Reply:
@mateuszmandera this is what folks in the [#backend > rate limiting](https://chat.zulip.org/#narrow/stream/3-backend/topic/rate.20limiting/near/828099) suggested. You can see the detailed discussion there.

Reply:
Your strategy of combining the entries from multiple rules into the same dictionary has invalidated the performance guarantees I proved about the GC algorithm. A short-duration rule may trigger frequent GCs that waste time scanning through lots of entries from a long-duration rule.